base_layer/common_types/src/tari_address/dual_address.rs

```diff
@@ -0,0 +1,418 @@
+            features,
+            public_view_key: view_key,
+            public_spend_key: spend_key,
+        }
+    }
+
+    /// Creates a new Tari Address from the provided public keys and network while using the default features
+    pub fn new_with_default_features(view_key: PublicKey, spend_key: PublicKey, network: Network) -> DualAddress {
+        Self {
+            network,
+            features: TariAddressFeatures::default(),
+            public_view_key: view_key,
+            public_spend_key: spend_key,
+        }
+    }
+
+    /// helper function to convert emojis to u8
+    pub fn emoji_to_bytes(emoji: &str) -> Result<Vec<u8>, TariAddressError> {
+        // The string must be the correct size, including the checksum
+        if emoji.chars().count() != INTERNAL_DUAL_SIZE {
+            return Err(TariAddressError::InvalidSize);
+        }
+
+        // Convert the emoji string to a byte array
+        let mut bytes = Vec::<u8>::with_capacity(INTERNAL_DUAL_SIZE);
+        for c in emoji.chars() {
+            if let Some(i) = REVERSE_EMOJI.get(&c) {
+                bytes.push(*i);
+            } else {
+                return Err(TariAddressError::InvalidEmoji);
+            }
+        }
+        Ok(bytes)
+    }
+
+    /// Construct an TariAddress from an emoji string with checksum trying to calculate the network
+    pub fn from_emoji_string(emoji: &str) -> Result<Self, TariAddressError> {
+        let bytes = Self::emoji_to_bytes(emoji)?;
+
+        Self::from_bytes(&bytes)
+    }
+
+    /// Gets the network from the Tari Address
+    pub fn network(&self) -> Network {
+        self.network
+    }
+
+    /// Gets the features from the Tari Address
+    pub fn features(&self) -> TariAddressFeatures {
+        self.features
+    }
+
+    /// Convert Tari Address to an emoji string with checksum
+    pub fn to_emoji_string(&self) -> String {
+        // Convert the public key to bytes and compute the checksum
+        let bytes = self.to_bytes();
+        bytes.iter().map(|b| EMOJI[*b as usize]).collect::<String>()
+    }
+
+    /// Return the public view key of an Tari Address
+    pub fn public_view_key(&self) -> &PublicKey {
+        &self.public_view_key
+    }
+
+    /// Return the public spend key of an Tari Address
+    pub fn public_spend_key(&self) -> &PublicKey {
+        &self.public_spend_key
+    }
+
+    /// Construct Tari Address from bytes and try to calculate the network
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, TariAddressError>
+    where Self: Sized {
+        if bytes.len() != INTERNAL_DUAL_SIZE {
+            return Err(TariAddressError::InvalidSize);
+        }
+        if validate_checksum(bytes).is_err() {
+            return Err(TariAddressError::InvalidChecksum);
+        }
+        let network = Network::try_from(bytes[0]).map_err(|_| TariAddressError::InvalidNetwork)?;
+        let features = TariAddressFeatures::from_bits(bytes[1]).ok_or(TariAddressError::InvalidFeatures)?;
```

Comment:
Dual address with interactive only features should not be allowed.
```suggestion
        let features = TariAddressFeatures::from_bits(bytes[1]).ok_or(TariAddressError::InvalidFeatures)?;
        if features == TariAddressFeatures::create_interactive_only() {
            return Err(TariAddressError::InvalidFeatures);
        }
```